ch_pipeline/hfb/analysis.py

```diff
@@ -351,6 +351,104 @@ def process(self, minuend, subtrahend):
         return out
 
 
+class HFBOnOffDifference(task.SingleTask):
+    """Computes the on-off differencing
+
+    Used for flattening sub-frequency band shape by differencing on-source and
+    off-source data.
+    """
+
+    offset = config.Property(proptype=int, default=5)
+    nsamples = config.Property(proptype=int, default=5)
+
+    def process(self, stream):
+        """Take difference and place in container.
+
+        Parameters
+        ----------
+        stream : HFBHighResRingMap
+            Container with stacked HFB data and weights.
+
+        Returns
+        -------
+        out : HFBHighResRingMap
+            Container with HFB data and weights; the result of the on-off differencing.
+        """
+
+        # Load data and weights
+        data = stream.hfb[:]
+        weight = stream.weight[:]
+        ra = stream.ra[:]
+        nra = len(ra)
+        # Create a 1D kernel to select off-source data.
+        kernel = np.zeros_like(ra)
+        # Fill desired elements with one. nsample elements on either sides
+        left = np.arange(self.offset + 1, self.offset + self.nsamples + 1)
+        right = np.arange(nra - self.offset - self.nsamples, nra - self.offset)
+        kernel[left] = 1
+        kernel[right] = 1
+        # We want the weighted average of off-source data, which is
+        # sum(off-source data * weights) / sum(off-source weights)
+
+        # Take the weighted sum of off-source data (numerator of the
+        # weighted average). To do that, convolve the kernel above
+        # with the data
+        ker_fft = np.fft.fft(kernel)
+        dat_fft = np.fft.fft(data * weight, axis=2)
+        sum_data = np.fft.ifft(
+            dat_fft * ker_fft[np.newaxis, np.newaxis, :, np.newaxis], axis=2
+        ).real
+
+        # Then, convolve the kernel above  with the weights to find the sum
+        # of off-source weights (denominator of the weighted average)
+        weight_fft = np.fft.fft(weight, axis=2)
+
+        # n is the sum of off-source weights over ra axis
+        n = np.fft.ifft(
```

Comment:
Not sure, but maybe it would be nice to call this variable `sum_weight`, since it's quite similar to `sum_data` above? Maybe I don't understand it well enough though